Add Storage Foundation article

tomayac opened the pull request:
Fixes #4487 

Changes proposed in this pull request:

- Add the article

tomayac commented:
@jpmedley You can start your review now. Thank you!

tomayac commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,372 @@
+// `["sunset", "sunrise", "noon"]`
+await storageFoundation.getAll();
+// Delete one of the three files.
+await storageFoundation.delete('noon');
```
@fivedots This example from the explainer actually doesn't work. It fails because the file is open. See https://github.com/WICG/storage-foundation-api-explainer/issues/24 for background.

petele reviewed (commented):
Looks good, a few minor comments inline. 

One general comment - in a few places you talk about what the API "aims" to do, or what it's goal is. For things like this, I recommend more concrete language - "the API let's developers do (something)".  The one time when hedging is good, is when you can't make a promise, especially around security or privacy.

jpmedley commented:
Do you mind clearing Pete's feedback first.

tomayac commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,372 @@
+The Storage Foundation API introduces a new interface `storageFoundation` that lives on the `window`
+object and that includes a number of functions:
+
+{% Aside %} We are currently exploring the tradeoffs between providing a synchronous vs. an
```
I don't think there is a way, but the good thing is that this Aside will eventually go away since it's temporary. There are one or two other such temporary Asides…

tomayac commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,372 @@
+{% Aside %} A file can only be opened once. This means concurrent access from different tabs is
+currently not possible. {% endAside %}
+
+- `storageFoundation.delete(name)`: Removes the file with the given name. Void.
```
Thanks. It returns `undefined`, which is what I meant with void. I'll just make it clear.

tomayac commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,372 @@
+{% Aside %} A file can only be opened once. This means concurrent access from different tabs is
+currently not possible. {% endAside %}
+
+- `storageFoundation.delete(name)`: Removes the file with the given name. Void.
```
```suggestion
- `storageFoundation.delete(name)`: Removes the file with the given name. Returns `undefined`.
```

tomayac commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,372 @@
+
+- `storageFoundation.delete(name)`: Removes the file with the given name. Void.
+- `storageFoundation.rename(oldName, newName)`: Renames the file from the old name to the new name
+  atomically. Void.
```
```suggestion
  atomically.  Returns `undefined`.
```

tomayac commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,372 @@
+{% Aside %} Storage Foundation API used to be called NativeIO. Some references to this name still
+remain and will be removed eventually. {% endAside %}
+
+- `NativeIOFile.close()`: Closes a file. Void.
```
```suggestion
- `NativeIOFile.close()`: Closes a file.  Returns `undefined`.
```

tomayac commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,372 @@
+
+- `NativeIOFile.close()`: Closes a file. Void.
+- `NativeIOFile.flush()`: Synchronizes (that is, flushes) a file's in-core state with the storage
+  device. Void.
```
```suggestion
  device.  Returns `undefined`.
```

tomayac commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,372 @@
+- `NativeIOFile.getLength()`: Returns a promise that resolves with the length of the file in bytes.
+- `NativeIOFile.setLength(length)`: Sets the length of the file in bytes. If the new length is
+  smaller than the current length, bytes are removed starting from the end of the file. Otherwise
+  the file is extended with zero-valued bytes. Void.
```
```suggestion
  the file is extended with zero-valued bytes.  Returns `undefined`.
```

tomayac commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,373 @@
+
+The Storage Foundation API was specified and implemented by
+[Emanuel Krivoy](https://github.com/fivedots) and [Richard Stotz](https://github.com/rstz). This
+article was reviewed by [Joe Medley](https://github.com/jpmedley).
```
```suggestion
article was reviewed by [Pete LePage](https://github.com/petele) and
[Joe Medley](https://github.com/jpmedley).
```

tomayac commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+  primitives upon which they can build higher-level components.
+authors:
+  - thomassteiner
+date: 2021-06-09
```
```suggestion
date: 2021-06-10
```

tomayac commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+interface, developers will be able to "bring their own storage" to the web, reducing the feature gap
+between web and platform-specific code.
+
+The Storage Foundation API is designed to resemble a very basic filesystem, aiming to give
```
```suggestion
The Storage Foundation API is designed to resemble a very basic filesystem so it gives
```

tomayac commented:
Thanks, @petele, for the suggestions. Applied all of them and credited you as a reviewer :-) @jpmedley, if you have time, I'd appreciate your go next. Thanks as always!

jpmedley commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+
+- Productivity or creativity apps that operate on large amounts of video, audio, or image data. Such
+  apps can offload segments to disk instead of holding them in memory.
+- Apps that rely on a persistent filesystem accessible from Wasm and that need more performance than
```
```suggestion
- Apps that rely on a persistent file system accessible from Wasm and that need more performance than
```

jpmedley commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+
+### File system calls
+
+The Storage Foundation API introduces a new interface `storageFoundation` that lives on the `window`
```
```suggestion
The Storage Foundation API introduces a new interface, `storageFoundation` that lives on the `window`
```

jpmedley commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+The Storage Foundation API introduces a new interface `storageFoundation` that lives on the `window`
+object and that includes a number of functions:
+
+{% Aside %} We are currently exploring the tradeoffs between providing a synchronous vs. an
```
```suggestion
{% Aside %} We are currently exploring the tradeoffs between providing a synchronous versus an
```

jpmedley commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+
+### File system calls
+
+The Storage Foundation API introduces a new interface `storageFoundation` that lives on the `window`
```
The interface is actually called `NativeIOFileManager`.

tomayac replied:
Correct. I reword to "object" now, since the `Native*` interface name will disappear. The object's name is final.

jpmedley commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+{% Aside %} A file can only be opened once. This means concurrent access from different tabs is
+currently not possible. {% endAside %}
+
+- `storageFoundation.delete(name)`: Removes the file with the given name. Returns `undefined`.
```
```suggestion
- `storageFoundation.delete(name)`: Removes the file with the given name. Returns a promise
  that resolves when the file is deleted.
```

tomayac replied:
Good catch!

jpmedley commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+
+- `storageFoundation.delete(name)`: Removes the file with the given name. Returns `undefined`.
+- `storageFoundation.rename(oldName, newName)`: Renames the file from the old name to the new name
+  atomically.  Returns `undefined`.
```
```suggestion
  atomically.  Returns `undefined`.
```
```suggestion
  atomically.  Returns a promise that resolves when the file is renamed.
```

jpmedley commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+{% Aside %} The Storage Foundation API achieves fast and predictable performance by implementing its
+own quota management system. Web applications must explicitly ask for capacity before storing any
+new data. This request will be granted according to the browser's quota guidelines. Anytime an
+application starts a new JavaScript execution context (e.g., new tab, new worker, reloading the
```
```suggestion
application starts a new JavaScript execution context (e.g., a new tab, a new worker, or reloading the
```

jpmedley commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+own quota management system. Web applications must explicitly ask for capacity before storing any
+new data. This request will be granted according to the browser's quota guidelines. Anytime an
+application starts a new JavaScript execution context (e.g., new tab, new worker, reloading the
+page), it must make sure it owns sufficient capacity before writing new data. {% endAside %}
```
>new data

Why not just 'data'? Is there some distinction here I'm missing?

tomayac replied:
Right, "new" should be replaced with "any".

jpmedley commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+application starts a new JavaScript execution context (e.g., new tab, new worker, reloading the
+page), it must make sure it owns sufficient capacity before writing new data. {% endAside %}
+
+- `storageFoundation.releaseCapacity(toBeReleasedCapacity)`: Releases unused capacity (in bytes)
```
Where's `requestCapacitySync()` and `releaseCapacitySync()`?

tomayac replied:
For now all functions are asynchronous and the synchronous functions have been removed.

jpmedley commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+application starts a new JavaScript execution context (e.g., new tab, new worker, reloading the
+page), it must make sure it owns sufficient capacity before writing new data. {% endAside %}
+
+- `storageFoundation.releaseCapacity(toBeReleasedCapacity)`: Releases unused capacity (in bytes)
```
```suggestion
- `storageFoundation.releaseCapacity(toBeReleasedCapacity)`: Releases the specified number of bytes
```

jpmedley replied:
I know that breaks your parallelism, but the grammar doesn't quite work for this definition.

jpmedley commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+page), it must make sure it owns sufficient capacity before writing new data. {% endAside %}
+
+- `storageFoundation.releaseCapacity(toBeReleasedCapacity)`: Releases unused capacity (in bytes)
+  from the current execution context. Returns a promised that resolves with the remaining amount of
+  capacity available.
```
```suggestion
  from the current execution context, and returns a promised that resolves with the remaining
  capacity.
```

jpmedley commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+
+Working with files happens via the following functions:
+
+{% Aside %} Storage Foundation API used to be called NativeIO. Some references to this name still
```
```suggestion
{% Aside %} The Storage Foundation API used to be called NativeIO. Some references to this name still
```

jpmedley replied:
You should move this up so that you can correctly name the interface returned by `window.storageFoundation`.

jpmedley commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+{% Aside %} Storage Foundation API used to be called NativeIO. Some references to this name still
+remain and will be removed eventually. {% endAside %}
+
+- `NativeIOFile.close()`: Closes a file.  Returns `undefined`.
```
```suggestion
- `NativeIOFile.close()`: Closes a file, and returns a promise that resolves when the operation completes.
```

jpmedley replied:
Please go through and clarify which operations return a promise.

tomayac replied:
Will do, thanks for catching those!

jpmedley commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+remain and will be removed eventually. {% endAside %}
+
+- `NativeIOFile.close()`: Closes a file.  Returns `undefined`.
+- `NativeIOFile.flush()`: Synchronizes (that is, flushes) a file's in-core state with the storage
```
What's an "in-core state"?

tomayac replied:
One optimization of the API is that it lies: it tells you something is persisted, when actually it's only stored in memory but not on disk yet.

jpmedley commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+  device.  Returns `undefined`.
+
+{% Aside %} It is a known issue that `flush()` might be slow and we are exploring whether offering a
+faster, less reliable variant would be useful. {% endAside %}
```
Since it returns a promise, why do I care? I don't always need to use `await`.

tomayac replied:
You're in trouble when you randomly want to access previously written bytes, but these bytes weren't flushed yet.

jpmedley commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+  smaller than the current length, bytes are removed starting from the end of the file. Otherwise
+  the file is extended with zero-valued bytes.  Returns `undefined`.
+- `NativeIOFile.read(buffer, offset)`: Reads the contents of the file at the given offset through a
+  buffer that is the result of transferring the given buffer, which is then left detached. Returns a
```
I'm not sure I follow this. Are buffered file operations beyond the scope of this article? Maybe link to a discussion and call out that there's a deeper story here.

What does 'detached' mean? Does it mean that I can't write to the buffer after I've called `read()`?

tomayac replied:
See the example:

```js
const readBuffer = new Uint8Array(3);
// Read at offset 1. `result.buffer` contains the transferred buffer,
// `result.readBytes` is 2, the number of bytes read. `readBuffer` is left
// detached.
result = await file.read(readBuffer, 1); 
// `Uint8Array(3) [65, 66, 0]`
```

It's a bit like an empty basket (`readBuffer`) that you take at the entrance of a store, put your groceries in, but then at the cashier your stuff gets moved to a bag (`result`) for you to take home.

jpmedley commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+  A `NativeIOReadResult` is an object that consists of two entries:
+
+  - `buffer`: An
+    [`ArrayBufferView`](https://developer.mozilla.org/en-US/docs/Web/API/ArrayBufferView) which is
```
```suggestion
    [`ArrayBufferView`](https://developer.mozilla.org/en-US/docs/Web/API/ArrayBufferView), which is
```

jpmedley commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+- `NativeIOFile.write(buffer, offset)`: Writes the contents of the given buffer into the file at the
+  given offset. The buffer is transferred before any data is written and is therefore left detached.
+  Returns a `NativeIOWriteResult` with the transferred buffer and the number of bytes that were
+  successfully written. The file will be extended if the write range spans beyond its length.
```
```suggestion
  successfully written. The file will be extended if the write range exceeds its length.
```

jpmedley commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+
+### Enabling support during the origin trial phase
+
+Starting in Chromium&nbsp;90, the Storage Foundation API will be available as an origin trial in
```
```suggestion
Starting in Chromium&nbsp;90, the Storage Foundation API is available as an origin trial in
```

jpmedley commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+defined in [Controlling Access to Powerful Web Platform Features][powerful-apis], including user
+control, transparency, and ergonomics.
+
+Following the same pattern as other modern storage web APIs, access to the Storage Foundation API is
```
```suggestion
Following the same pattern as other modern web storage APIs, access to the Storage Foundation API is
```

tomayac replied:
In this case I think it's implied by the topic, and making it explicit may point people in the direction of just web storage as in [Web Storage](https://developer.mozilla.org/en-US/docs/Web/API/Web_Storage_API). It's all storage that happens on the web, including IDB and friends.

jpmedley commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+
+### User control
+
+Storage quota will be used to distribute access to disk space and to prevent abuse. Like other
```
I'm not sure what you mean.

tomayac replied:
Memory you want to occupy needs to be requested first.

jpmedley commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+### User control
+
+Storage quota will be used to distribute access to disk space and to prevent abuse. Like other
+storage APIs, users have the option of clearing the space taken by Storage Foundation API through
```
```suggestion
storage APIs, users can clear the space taken by Storage Foundation API through
```

tomayac commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+layout: post
+title: 'High performance storage for your app: the Storage Foundation API'
+subhead: |
+  The Storage Foundation API resembles a very basic filesystem,
```
OK, will do!

tomayac commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+  with direct access to stored data through buffers and offsets. It gives
+  developers flexibility by providing generic, simple, and performant primitives upon
+  which they can build higher-level components. It is particularly well suited for
+  Wasm-based libraries and applications that want to use custom storage algorithms to
```
```suggestion
  Wasm-based libraries and applications that perform best when they can use custom storage algorithms to
```

tomayac commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+  with direct access to stored data through buffers and offsets. It gives
+  developers flexibility by providing generic, simple, and performant primitives upon
+  which they can build higher-level components. It is particularly well suited for
+  Wasm-based libraries and applications that want to use custom storage algorithms to
```
Now using "applications that perform best when they can use…"

tomayac commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+high-performance applications for the web. Most notably,
+[WebAssembly](https://developer.mozilla.org/en-US/docs/WebAssembly) (Wasm) has opened the door to
+fast and powerful web applications, while technologies like [Emscripten](https://emscripten.org/)
+now allow developers to reuse tried and tested code on the web. In order to truly leverage this
```
At this stage I should really use a lint rule to simplify "in order to" to just "to"…

tomayac commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+potential, developers must be given the same power and flexibility when it comes to storage.
+
+This is where the Storage Foundation API comes in. The Storage Foundation API is a new fast and
+unopinionated storage API that unlocks new and much-requested use cases for the web, such as
```
Not enforcing any opinionated approaches to how storage is done.

tomayac commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,374 @@
+own quota management system. Web applications must explicitly ask for capacity before storing any
+new data. This request will be granted according to the browser's quota guidelines. Anytime an
+application starts a new JavaScript execution context (e.g., new tab, new worker, reloading the
+page), it must make sure it owns sufficient capacity before writing new data. {% endAside %}
```
```suggestion
page), it must make sure it owns sufficient capacity before writing any data. {% endAside %}
```

tomayac commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,375 @@
+remain and will be removed eventually. {% endAside %}
+
+- `NativeIOFile.close()`: Closes a file, and returns a promise that resolves when the operation completes.
+- `NativeIOFile.flush()`: Synchronizes (that is, flushes) a file's in-core state with the storage
```
```suggestion
- `NativeIOFile.flush()`: Synchronizes (that is, flushes) a file's in-memory state with the storage
```

tomayac commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,375 @@
+
+### User control
+
+Storage quota will be used to distribute access to disk space and to prevent abuse. Like other
```
```suggestion
Storage quota will be used to distribute access to disk space and to prevent abuse. Memory you want to occupy needs to be requested first. Like other
```

tomayac commented:
Thanks, @jpmedley. Addressed your comments and remarks, and also verified all return types (I misread `Promise<void>` as just `void`). It's corrected throughout now.

tomayac commented on src/site/content/en/blog/storage-foundation/index.md:
```diff
@@ -0,0 +1,378 @@
+  primitives on which they can build higher-level components.
+authors:
+  - thomassteiner
+date: 2021-06-11
```
```suggestion
date: 2021-06-15
```
